File: android/app/src/test/kotlin/com/simplephotos/data/media/RenditionChoiceTest.kt
package com.simplephotos.data.media

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * Mirrors `web/src/gallery/renditionChoice.test.ts` so both clients agree on
 * what the picker offers and what it defaults to (#49). Where the two diverge
 * it is because Android can answer "is this link expensive?" definitively and
 * web cannot — see [isConstrained] versus web's three-signal guess.
 *
 * The live-library shapes below are the measured census from `todo.md`, not
 * invented examples: 71 of 742 videos are portrait, 14 of them exactly
 * `1080x1920`, and 4 are `2288x1088`. Both groups break a naive height-based
 * rung rule, which is why the ladder keys on the short edge.
 */
class RenditionChoiceTest {

    /** Landscape 16:9 unless a test says otherwise. */
    private fun rung(
        shortEdge: Int,
        isSource: Boolean = false,
        blobId: String? = "b-$shortEdge",
        width: Int = Math.round(shortEdge * 16f / 9f),
        height: Int = shortEdge,
    ) = Rendition(
        shortEdge = shortEdge,
        width = width,
        height = height,
        isSource = isSource,
        blobId = blobId,
        codec = "h264",
        sizeBytes = shortEdge * 1000L,
    )

    // ── offerableRenditions ─────────────────────────────────────────────────

    @Test
    fun offerable_isEmptyForTheNormalCaseOfOneQuality() {
        assertEquals(emptyList<Rendition>(), offerableRenditions(emptyList()))
        assertEquals(emptyList<Rendition>(), offerableRenditions(null))
    }

    @Test
    fun offerable_sortsHighestFirstRatherThanTrustingServerOrder() {
        val got = offerableRenditions(listOf(rung(1080), rung(2160, isSource = true)))
        assertEquals(listOf(2160, 1080), got.map { it.shortEdge })
    }

    @Test
    fun offerable_neverOffersTheSameQualityTwice() {
        val got = offerableRenditions(listOf(rung(1080), rung(1080), rung(2160)))
        assertEquals(listOf(2160, 1080), got.map { it.shortEdge })
    }

    @Test
    fun offerable_dropsRungsThisPlayerCannotStream() {
        // Null blobId is an unencrypted install. This player only speaks
        // spblob://, so such a rung would be a menu entry that does nothing.
        val got = offerableRenditions(listOf(rung(2160, isSource = true), rung(1080, blobId = null)))
        assertEquals(listOf(2160), got.map { it.shortEdge })
    }

    // ── shouldOfferPicker ───────────────────────────────────────────────────

    @Test
    fun picker_notDrawnWhenThereIsNothingToChooseBetween() {
        assertFalse(shouldOfferPicker(emptyList()))
        assertFalse(shouldOfferPicker(null))
    }

    @Test
    fun picker_notDrawnForALoneSourceRung() {
        // A one-entry menu implies a choice the user does not have.
        assertFalse(shouldOfferPicker(listOf(rung(2160, isSource = true))))
    }

    @Test
    fun picker_notDrawnWhenTheOnlyAlternativeIsUnstreamable() {
        // Two rungs on the wire, but one is plaintext-only — so after filtering
        // there is again nothing to choose between. Counting before filtering
        // would draw a gear icon opening a one-entry menu.
        assertFalse(
            shouldOfferPicker(listOf(rung(2160, isSource = true), rung(1080, blobId = null)))
        )
    }

    @Test
    fun picker_drawnOnceARealAlternativeExists() {
        assertTrue(shouldOfferPicker(listOf(rung(2160, isSource = true), rung(1080))))
    }

    // ── isConstrained ───────────────────────────────────────────────────────

    @Test
    fun constrained_meteredAloneNeverDowngrades() {
        // The issue is explicit: with the data saver OFF, always serve highest
        // regardless of network. A metered link must not downgrade a user who
        // never asked for it.
        assertFalse(isConstrained(dataSaverEnabled = false, metered = true))
    }

    @Test
    fun constrained_dataSaverAloneNeverDowngrades() {
        assertFalse(isConstrained(dataSaverEnabled = true, metered = false))
    }

    @Test
    fun constrained_requiresBoth() {
        assertTrue(isConstrained(dataSaverEnabled = true, metered = true))
        assertFalse(isConstrained(dataSaverEnabled = false, metered = false))
    }

    // ── chooseDefaultRendition ──────────────────────────────────────────────

    @Test
    fun default_isNullWhenThereIsNoLadder() {
        // Not an error — it means "play the photo's own blob as before #49".
        assertNull(chooseDefaultRendition(emptyList(), constrained = false))
        assertNull(chooseDefaultRendition(null, constrained = true))
    }

    @Test
    fun default_isHighestOnAnUnconstrainedLink() {
        val got = chooseDefaultRendition(listOf(rung(1080), rung(2160, isSource = true)), false)
        assertEquals(2160, got?.shortEdge)
    }

    @Test
    fun default_capsAtAbsolute1080NotOneRungDown() {
        // The 8K case from the live census. One rung down from 7680x4320 is 4K,
        // which is not what "lower on cellular" means to anybody — so the cap is
        // an absolute quality, not a relative step.
        val ladder = listOf(rung(4320, isSource = true), rung(2160), rung(1080))
        val got = chooseDefaultRendition(ladder, constrained = true)
        assertEquals(CONSTRAINED_MAX_SHORT_EDGE, got?.shortEdge)
    }

    @Test
    fun default_takesSmallestWhenEveryRungExceedsTheCap() {
        // A 4K source whose 1080 rung has not been generated yet. Refusing to
        // play would be worse: the alternative is a metered client fetching 4K.
        val ladder = listOf(rung(4320, isSource = true), rung(2160))
        val got = chooseDefaultRendition(ladder, constrained = true)
        assertEquals(2160, got?.shortEdge)
    }

    @Test
    fun default_neverPicksARungThePickerRefusesToShow() {
        // Defaulting to a filtered-out rung would leave the menu ticking nothing
        // and no way back to it.
        val ladder = listOf(rung(2160, isSource = true), rung(1080, blobId = null))
        val got = chooseDefaultRendition(ladder, constrained = true)
        assertEquals(2160, got?.shortEdge)
    }

    // ── the live library's awkward shapes ───────────────────────────────────

    @Test
    fun portrait1080x1920IsThe1080pTierNotA1920pOne() {
        // 14 live videos are exactly this. Keyed on height they would look like
        // a 1920p source needing a downscale; keyed on the short edge they are
        // already the 1080p tier and need no rung at all.
        val r = rung(1080, isSource = true, width = 1080, height = 1920)
        assertEquals("Original (1080p)", formatRenditionLabel(r))
        assertFalse(shouldOfferPicker(listOf(r)))
    }

    @Test
    fun portraitAndLandscapeOfTheSameTierSortTogether() {
        // A portrait 1080x1920 rung and a landscape 3840x2160 source: the
        // ordering must come from the short edge, not from either raw dimension
        // (1920 > 2160 is false, but 1920 > 1080 is true — pick the wrong field
        // and the portrait rung sorts above the 4K original).
        val ladder = listOf(
            rung(1080, width = 1080, height = 1920),
            rung(2160, isSource = true, width = 3840, height = 2160),
        )
        assertEquals(listOf(2160, 1080), offerableRenditions(ladder).map { it.shortEdge })
    }

    // ── labels ──────────────────────────────────────────────────────────────

    @Test
    fun label_showsTheResolutionEvenForTheOriginal() {
        // "Original" alone forces the user to guess whether it is bigger than
        // the 1080p entry below it.
        assertEquals("Original (2160p)", formatRenditionLabel(rung(2160, isSource = true)))
        assertEquals("1080p", formatRenditionLabel(rung(1080)))
    }

    // ── renditionsEqual ─────────────────────────────────────────────────────

    @Test
    fun equal_treatsNullAndEmptyAsTheSameState() {
        // Both mean "one quality" and they arrive from different places: a
        // pre-#49 server yields null, a #49 server sends an empty list for the
        // ~600 videos needing no rung. Treating them as different makes the
        // first pass after a server upgrade rewrite the whole library.
        assertTrue(renditionsEqual(null, emptyList()))
        assertTrue(renditionsEqual(emptyList(), null))
        assertTrue(renditionsEqual(null, null))
    }

    @Test
    fun equal_detectsARealLadderChange() {
        val before = listOf(rung(2160, isSource = true))
        val after = listOf(rung(2160, isSource = true), rung(1080))
        assertFalse(renditionsEqual(before, after))
    }

    @Test
    fun equal_detectsARungGainingItsBlob() {
        // The moment a rung becomes playable is an UPDATE, not an INSERT — the
        // trigger bug `036` had to fix. The client must notice it too.
        assertFalse(
            renditionsEqual(listOf(rung(1080, blobId = null)), listOf(rung(1080)))
        )
    }
}

File: web/src/gallery/renditionChoice.test.ts
import { describe, expect, it } from "vitest";
import {
  CONSTRAINED_MAX_SHORT_EDGE,
  chooseDefaultRendition,
  formatRenditionLabel,
  isConstrainedNetwork,
  offerableRenditions,
  renditionsEqual,
  shouldOfferPicker,
  type Rendition,
} from "./renditionChoice";

function rung(short_edge: number, over: Partial<Rendition> = {}): Rendition {
  return {
    short_edge,
    // Landscape 16:9 unless a test says otherwise.
    width: Math.round((short_edge * 16) / 9),
    height: short_edge,
    is_source: false,
    blob_id: `b-${short_edge}`,
    codec: "h264",
    size_bytes: short_edge * 1000,
    ...over,
  };
}

describe("offerableRenditions", () => {
  it("is empty for the normal case of a video with one quality", () => {
    expect(offerableRenditions([])).toEqual([]);
    expect(offerableRenditions(undefined)).toEqual([]);
  });

  it("sorts highest first even when the server's order is not trusted", () => {
    const got = offerableRenditions([rung(1080), rung(2160, { is_source: true })]);
    expect(got.map((r) => r.short_edge)).toEqual([2160, 1080]);
  });

  it("never offers the same quality twice", () => {
    const got = offerableRenditions([rung(1080), rung(1080), rung(2160)]);
    expect(got.map((r) => r.short_edge)).toEqual([2160, 1080]);
  });
});

describe("shouldOfferPicker", () => {
  it("draws no gear icon when there is nothing to choose between", () => {
    // The overwhelmingly common case: only videos above the 1080p tier ever
    // get a second rung, so most of the library lands here.
    expect(shouldOfferPicker([])).toBe(false);
    expect(shouldOfferPicker(undefined)).toBe(false);
  });

  it("draws no gear icon for a lone source rung", () => {
    // A one-entry menu implies a choice the user does not have.
    expect(shouldOfferPicker([rung(2160, { is_source: true })])).toBe(false);
  });

  it("draws the gear icon once a real alternative exists", () => {
    expect(shouldOfferPicker([rung(2160, { is_source: true }), rung(1080)])).toBe(true);
  });
});

describe("isConstrainedNetwork", () => {
  it("is false when the browser reports nothing", () => {
    // Safari and Firefox expose no `connection` at all. Defaulting to
    // "constrained" there would cap every desktop viewer at 1080p forever.
    expect(isConstrainedNetwork(undefined)).toBe(false);
    expect(isConstrainedNetwork({})).toBe(false);
  });

  it("honours an explicit Data Saver request over everything else", () => {
    expect(isConstrainedNetwork({ saveData: true, type: "wifi", effectiveType: "4g" })).toBe(
      true,
    );
  });

  it("treats a cellular link as constrained however fast it measures", () => {
    // The issue asks for wifi-vs-cellular, not fast-vs-slow. A 5G link reports
    // effectiveType "4g", so throughput alone would never see this case.
    expect(isConstrainedNetwork({ type: "cellular", effectiveType: "4g" })).toBe(true);
  });

  it("treats a slow link as constrained when the engine hides the link type", () => {
    expect(isConstrainedNetwork({ effectiveType: "2g" })).toBe(true);
    expect(isConstrainedNetwork({ effectiveType: "3g" })).toBe(true);
    expect(isConstrainedNetwork({ effectiveType: "4g" })).toBe(false);
  });

  it("does not constrain wifi", () => {
    expect(isConstrainedNetwork({ type: "wifi", effectiveType: "4g" })).toBe(false);
  });
});

describe("chooseDefaultRendition", () => {
  const ladder4k = [rung(2160, { is_source: true }), rung(1080)];

  it("returns nothing when there are no renditions, meaning 'play the photo's own blob'", () => {
    expect(chooseDefaultRendition([])).toBeUndefined();
    expect(chooseDefaultRendition(undefined)).toBeUndefined();
  });

  it("defaults to the original on an unmetered link", () => {
    expect(chooseDefaultRendition(ladder4k, { type: "wifi" })?.short_edge).toBe(2160);
    expect(chooseDefaultRendition(ladder4k)?.short_edge).toBe(2160);
  });

  it("defaults to 1080p on cellular", () => {
    expect(chooseDefaultRendition(ladder4k, { type: "cellular" })?.short_edge).toBe(1080);
  });

  it("caps at a quality, not at one rung down", () => {
    // An 8K source's next rung down is 4K. "Lower on cellular" does not mean
    // "4K instead of 8K" to anybody, which is why the cap is absolute.
    const ladder8k = [rung(4320, { is_source: true }), rung(2160), rung(1080)];
    expect(chooseDefaultRendition(ladder8k, { saveData: true })?.short_edge).toBe(
      CONSTRAINED_MAX_SHORT_EDGE,
    );
  });

  it("falls back to the smallest rung when nothing sits under the cap", () => {
    // A 4K source whose 1080 rung has not finished encoding yet. Refusing to
    // choose would leave a metered client fetching the 4K anyway.
    const partial = [rung(2160, { is_source: true }), rung(1440)];
    expect(chooseDefaultRendition(partial, { type: "cellular" })?.short_edge).toBe(1440);
  });

  it("picks the highest rung, not the source rung, when they differ", () => {
    // `is_source` marks the original, which is normally also the largest — but
    // nothing in the schema guarantees it, and reading index 0 as "the source"
    // is the kind of assumption that survives until it doesn't.
    const odd = [rung(1080, { is_source: true }), rung(2160)];
    expect(chooseDefaultRendition(odd, { type: "wifi" })?.short_edge).toBe(2160);
  });
});

describe("renditionsEqual", () => {
  it("treats undefined and empty as the same state", () => {
    // Not pedantry. A pre-#49 server (and every row cached before the field
    // existed) yields undefined; a #49 server sends [] for the ~600 videos that
    // need no rung. Calling those different rewrites the whole library on the
    // first pass after an upgrade, and every video on every pass after that.
    expect(renditionsEqual(undefined, [])).toBe(true);
    expect(renditionsEqual([], undefined)).toBe(true);
    expect(renditionsEqual(undefined, undefined)).toBe(true);
  });

  it("sees a ladder appearing", () => {
    expect(renditionsEqual([], [rung(2160, { is_source: true }), rung(1080)])).toBe(false);
  });

  it("sees a ladder being withdrawn", () => {
    expect(renditionsEqual([rung(1080)], [])).toBe(false);
  });

  it("sees a rung's blob being replaced by a re-encode", () => {
    // `upsert_rendition` refreshes a rung in place, so a re-encode changes the
    // blob id under a short_edge that did not move. Comparing lengths or
    // short_edges alone would miss it and leave the viewer fetching bytes the
    // server has already replaced.
    expect(renditionsEqual([rung(1080)], [rung(1080, { blob_id: "b-new" })])).toBe(false);
  });

  it("sees a rung's size changing", () => {
    expect(renditionsEqual([rung(1080)], [rung(1080, { size_bytes: 99 })])).toBe(false);
  });

  it("reports no change for an identical ladder", () => {
    const ladder = [rung(2160, { is_source: true }), rung(1080)];
    expect(renditionsEqual(ladder, [rung(2160, { is_source: true }), rung(1080)])).toBe(true);
  });
});

describe("formatRenditionLabel", () => {
  it("names the original and still shows its resolution", () => {
    expect(formatRenditionLabel(rung(2160, { is_source: true }))).toBe("Original (2160p)");
  });

  it("labels a rung by its short edge", () => {
    expect(formatRenditionLabel(rung(1080))).toBe("1080p");
  });

  it("labels a portrait video by its short edge, not its height", () => {
    // 14 live videos are 1080x1920. Labelling by height would call this
    // "1920p" — a resolution that does not exist — and the whole ladder keys
    // on the short edge precisely so this case is not special.
    const portrait = rung(1080, { width: 1080, height: 1920, is_source: true });
    expect(formatRenditionLabel(portrait)).toBe("Original (1080p)");
  });
});
